File: internal/content/conditions_fixture_test.go
// SPDX-License-Identifier: Apache-2.0

package content

import (
	"encoding/json"
	"os"
	"path/filepath"
	"slices"
	"testing"
)

// fixtureField is one field as the shared conditions fixture declares it.
type fixtureField struct {
	Key        string `json:"key"`
	Kind       string `json:"kind"`
	Multiple   bool   `json:"multiple"`
	Conditions []any  `json:"conditions"`
}

// conditionsFixture is the shared table both evaluators must pass.
type conditionsFixture struct {
	Operators []struct {
		Kind     string   `json:"kind"`
		Multiple bool     `json:"multiple"`
		Offers   []string `json:"offers"`
	} `json:"operators"`
	Compare []struct {
		Operator string `json:"operator"`
		Held     any    `json:"held"`
		Value    string `json:"value"`
		Holds    bool   `json:"holds"`
	} `json:"compare"`
	Hidden []struct {
		Name   string         `json:"name"`
		Fields []fixtureField `json:"fields"`
		Scope  map[string]any `json:"scope"`
		Hidden []string       `json:"hidden"`
	} `json:"hidden"`
}

// loadConditionsFixture reads the shared fixture from testdata.
func loadConditionsFixture(t *testing.T) conditionsFixture {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("testdata", "conditions.json"))
	if err != nil {
		t.Fatalf("reading the fixture: %v", err)
	}
	var fixture conditionsFixture
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("decoding the fixture: %v", err)
	}
	return fixture
}

// fieldOf returns the field a fixture entry declares.
func declaredField(declared fixtureField) Field {
	settings := map[string]any{}
	if declared.Multiple {
		settings[SettingMultiple] = true
	}
	if declared.Conditions != nil {
		settings[SettingConditions] = declared.Conditions
	}
	return Field{Key: declared.Key, Kind: FieldKind(declared.Kind), Settings: settings}
}

func TestFixtureOperatorsPerKind(t *testing.T) {
	t.Parallel()

	for _, row := range loadConditionsFixture(t).Operators {
		offered := SourceOperators(FieldKind(row.Kind), row.Multiple)
		if !slices.Equal(offered, row.Offers) && (len(offered) > 0 || len(row.Offers) > 0) {
			t.Errorf("SourceOperators(%s, %v) = %v, want %v", row.Kind, row.Multiple, offered, row.Offers)
		}
	}
}

func TestFixtureCompareTable(t *testing.T) {
	t.Parallel()

	for _, row := range loadConditionsFixture(t).Compare {
		if got := compare(row.Operator, row.Held, row.Value); got != row.Holds {
			t.Errorf("compare(%q, %v, %q) = %v, want %v", row.Operator, row.Held, row.Value, got, row.Holds)
		}
	}
}

func TestFixtureHiddenScopes(t *testing.T) {
	t.Parallel()

	for _, scenario := range loadConditionsFixture(t).Hidden {
		fields := make([]Field, len(scenario.Fields))
		for i, declared := range scenario.Fields {
			fields[i] = declaredField(declared)
		}

		hidden := Hidden(fields, scenario.Scope)

		got := make([]string, 0, len(hidden))
		for key := range hidden {
			got = append(got, key)
		}
		slices.Sort(got)
		want := slices.Clone(scenario.Hidden)
		slices.Sort(want)
		if !slices.Equal(got, want) {
			t.Errorf("%s: Hidden() = %v, want %v", scenario.Name, got, want)
		}
	}
}

File: frontend/src/test/conditions-fixture.test.ts
// SPDX-License-Identifier: Apache-2.0

import { readFileSync } from 'node:fs'
import { join } from 'node:path'

import { expect, test } from 'vitest'

import { repositoryRoot } from '../../scripts/config.ts'
import { compare, hiddenKeys, operatorsFor } from '../content/conditions'
import type { ConditionField } from '../content/conditions'

/** One row of the operators the fixture says each kind offers. */
interface OperatorRow {
	kind: string
	multiple?: boolean
	offers: string[]
}

/** One row of the comparisons the fixture pins. */
interface CompareRow {
	operator: string
	held: unknown
	value: string
	holds: boolean
}

/** One scope the fixture says hides a known set of keys. */
interface HiddenRow {
	name: string
	fields: { key: string; kind: string; multiple?: boolean; conditions?: unknown }[]
	scope: Record<string, unknown>
	hidden: string[]
}

/** The shared table both evaluators answer to. */
interface Fixture {
	operators: OperatorRow[]
	compare: CompareRow[]
	hidden: HiddenRow[]
}

const FIXTURE = JSON.parse(
	readFileSync(join(repositoryRoot(), 'internal', 'content', 'testdata', 'conditions.json'), 'utf8'),
) as Fixture

/**
 * Returns the fixture's field as the evaluator reads one.
 * @param declared - The field the fixture declares.
 * @returns The field to evaluate.
 */
function fieldOf(declared: HiddenRow['fields'][number]): ConditionField {
	const settings: Record<string, unknown> = {}
	if (declared.multiple === true) {
		settings.multiple = true
	}
	if (declared.conditions !== undefined) {
		settings.conditions = declared.conditions
	}
	return { key: declared.key, kind: declared.kind, settings }
}

test('offers the operators the server offers for every kind', () => {
	for (const row of FIXTURE.operators) {
		expect([row.kind, row.multiple === true, operatorsFor(row.kind, row.multiple === true)]).toEqual([
			row.kind,
			row.multiple === true,
			row.offers,
		])
	}
})

test('compares a value the way the server compares it', () => {
	for (const row of FIXTURE.compare) {
		expect([row.operator, row.held, row.value, compare(row.operator, row.held, row.value)]).toEqual([
			row.operator,
			row.held,
			row.value,
			row.holds,
		])
	}
})

test('hides the keys the server hides on every scope', () => {
	for (const row of FIXTURE.hidden) {
		const held = [...hiddenKeys(row.fields.map(fieldOf), row.scope)].sort()
		expect([row.name, held]).toEqual([row.name, [...row.hidden].sort()])
	}
})
